docs/writing-rules/data-flow/constant-propagation.md

```diff
@@ -7,37 +7,88 @@ tags:
 
 For example:
 
 <iframe src="https://semgrep.dev/embed/editor?snippet=Dyzd" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-Mutable objects
----------------
+## Mutable objects
 
-In general, Semgrep assumes that constant objects are immutable and they will not be modified by function calls. This may lead to false positives, especially in languages where strings are mutable such as C and Ruby.
+In general, Semgrep assumes that constant objects are immutable and won't be modified by function calls. This may lead to false positives, especially in languages where strings are mutable such as C and Ruby.
 
-The only exceptions (for now) are method calls whose returning value is being ignored. In those cases, Semgrep assumes that the method call may be mutating the callee object. This helps reducing false positives in Ruby, for example:
+The only exceptions are method calls whose returning value is ignored. In these cases, Semgrep assumes that the method call may be mutating the callee object. This helps reducing false positives in Ruby. For example:
 
 <iframe src="https://semgrep.dev/embed/editor?snippet=08yB" border="0" frameBorder="0" width="100%" height="432"></iframe>
 
-
-Disabling constant propagation
-------------------------------
-
-It is possible to disable constant propagation in a per-rule basis via rule [`options:`](/writing-rules/rule-syntax/#options) by setting `constant_propagation: false`.
+If constant propagation doesn't seem to work, consider whether the constant may be unexpectedly mutable. For example, given the following rule designed to taint the `REGEX` class variable:
+
+```yaml
+rules:
+  - id: redos-detection
+    message: Potential ReDoS vulnerability detected with $REGEX
+    severity: ERROR
+    languages:
+      - java
+    mode: taint
+    options:
+      symbolic_propagation: true
+    pattern-sources:
+      - patterns:
+          - pattern: $REDOS
+          - metavariable-analysis:
+              analyzer: redos
+              metavariable: $REDOS
+    pattern-sinks:
+      - pattern: Pattern.compile(...)
+```
+
+Semgrep fails to match its use in `Test2` when presented with the following code:
+
+```java
+import java.util.regex.Pattern;
+
+public String REGEX = "(a+)+$";
+
+public class Test2 {
+    public static void main(String[] args) {
+        Pattern pattern = Pattern.compile(REGEX);
+    }
+}
+```
+
+However, if you change the variable from `public` to `private`, Semgrep does return a match:
+
+```java
+import java.util.regex.Pattern;
+
+private String REGEX = "(a+)+$";
+
+public class Test2 {
+    public static void main(String[] args) {
+        Pattern pattern = Pattern.compile(REGEX);
+    }
+}
+```
+
+Because `REGEX` is public in the first code snippet, Semgrep doesn't propagate its value to other classes on the assumption that it could have mutated. However, in the second example, `REGEX` is private, and Semgrep assumes it to be immutable. The rule would also work with:
```

Comment:
i think in this case, it's actually a two step process. semgrep sees that it's `private`, _and also_ knows its only assigned to once, and so assumes it to be immutable. this code doesn't work if you add another assignment into the target:
```
import java.util.regex.Pattern;

private String REGEX = "(a+)+$";
REGEX = "(a+)+$";

public class Test2 {
    public static void main(String[] args) {
        Pattern pattern = Pattern.compile(REGEX);
    }
}
```
however, if it were `final`, then it would work always